File: apps/host/internal/servicedef/logs.go
package servicedef

import (
	"bytes"
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

const (
	// DefaultLogLines is the tail size when the operator does not ask for one.
	DefaultLogLines = 200
	// MaxLogLines caps --lines. A diagnostic tail is for looking at a recent
	// failure, not for shipping the whole file through a terminal.
	MaxLogLines = 5000
	// maxTailBytes bounds how far back the reader seeks. A log that wrote one
	// enormous line cannot make this command allocate without limit.
	maxTailBytes = 4 << 20
	// tailChunkBytes is the backwards read granularity.
	tailChunkBytes = 64 << 10
)

// Tail returns the last `lines` lines of path, reading only the end of the
// file: it seeks backwards in chunks and stops as soon as it has enough
// newlines, so a multi-gigabyte log costs the same as a small one. The returned
// slice never exceeds `lines` entries and never more than maxTailBytes of text.
func Tail(path string, lines int) ([]string, error) {
	if lines <= 0 {
		lines = DefaultLogLines
	}
	if lines > MaxLogLines {
		lines = MaxLogLines
	}
	file, err := os.Open(path)
	if err != nil {
		return nil, err
	}
	defer file.Close()
	info, err := file.Stat()
	if err != nil {
		return nil, err
	}
	if !info.Mode().IsRegular() {
		return nil, fmt.Errorf("%w: log file must be a regular file", ErrInvalid)
	}
	size := info.Size()
	var collected []byte
	offset := size
	truncated := false
	for offset > 0 {
		if int64(len(collected)) >= maxTailBytes {
			truncated = true
			break
		}
		step := int64(tailChunkBytes)
		if offset < step {
			step = offset
		}
		offset -= step
		chunk := make([]byte, step)
		if _, err := file.ReadAt(chunk, offset); err != nil && !errors.Is(err, io.EOF) {
			return nil, err
		}
		collected = append(chunk, collected...)
		// One extra newline: the first line in the buffer may be a fragment of
		// an earlier line, and it is dropped below unless we reached the start.
		if bytes.Count(collected, []byte{'\n'}) > lines {
			break
		}
	}
	// The first buffered line is a fragment unless the read reached the start of
	// the file or stopped exactly after a newline.
	partial := truncated
	if offset > 0 && !partial {
		var previous [1]byte
		if _, err := file.ReadAt(previous[:], offset-1); err != nil && !errors.Is(err, io.EOF) {
			return nil, err
		}
		partial = previous[0] != '\n'
	}
	text := strings.TrimRight(string(collected), "\n")
	if text == "" {
		return []string{}, nil
	}
	split := strings.Split(text, "\n")
	if partial && len(split) > 1 {
		split = split[1:]
	}
	if len(split) > lines {
		split = split[len(split)-lines:]
	}
	for index, line := range split {
		split[index] = strings.TrimRight(line, "\r")
	}
	return split, nil
}

// LatestStartupLog returns the newest `startup-*.log` the background start
// command wrote into the data directory, or os.ErrNotExist when there is none.
// Names are compared as a tiebreaker so the choice is deterministic when two
// files share a modification time.
func LatestStartupLog(dataDir string) (string, error) {
	matches, err := filepath.Glob(filepath.Join(dataDir, "startup-*.log"))
	if err != nil {
		return "", err
	}
	sort.Strings(matches)
	newest := ""
	var newestTime int64
	for _, candidate := range matches {
		info, err := os.Lstat(candidate)
		if err != nil || !info.Mode().IsRegular() {
			continue
		}
		if stamp := info.ModTime().UnixNano(); newest == "" || stamp > newestTime {
			newest, newestTime = candidate, stamp
		}
	}
	if newest == "" {
		return "", fmt.Errorf("no startup log in %s: %w", dataDir, os.ErrNotExist)
	}
	return newest, nil
}

File: apps/server/src/service/logs.ts
import { closeSync, fstatSync, openSync, readSync } from "node:fs";
import { ServiceDefinitionError } from "./spec";

/**
 * 日志尾部。
 *
 * 从文件末尾往回按块读，够行就停：一份几个 GB 的日志和一份几 KB 的代价一样。
 * 两个上限都在——行数和字节数——因为「一行写了 200 MB」的日志不该让这条命令
 * 无限分配内存。
 */

export const DEFAULT_LINES = 200;
export const MAX_LINES = 5000;
const MAX_TAIL_BYTES = 4 << 20;
const CHUNK_BYTES = 64 << 10;

export function tail(path: string, lines = DEFAULT_LINES): string[] {
  const wanted = Math.min(
    Math.max(Number.isInteger(lines) && lines > 0 ? lines : DEFAULT_LINES, 1),
    MAX_LINES,
  );
  const handle = openSync(path, "r");
  try {
    const stat = fstatSync(handle);
    if (!stat.isFile()) {
      throw new ServiceDefinitionError("日志必须是一个普通文件");
    }
    let offset = stat.size;
    let collected = Buffer.alloc(0);
    let truncated = false;
    while (offset > 0) {
      if (collected.byteLength >= MAX_TAIL_BYTES) {
        truncated = true;
        break;
      }
      const step = Math.min(CHUNK_BYTES, offset);
      offset -= step;
      const chunk = Buffer.alloc(step);
      readSync(handle, chunk, 0, step, offset);
      collected = Buffer.concat([chunk, collected]);
      // 多读一行：缓冲区里的第一行可能是上一行的残段，除非读到了文件开头。
      if (count(collected, 0x0a) > wanted) break;
    }
    let partial = truncated;
    if (offset > 0 && !partial) {
      const previous = Buffer.alloc(1);
      readSync(handle, previous, 0, 1, offset - 1);
      partial = previous[0] !== 0x0a;
    }
    const text = collected.toString("utf8").replace(/\n+$/, "");
    if (text === "") return [];
    let split = text.split("\n");
    if (partial && split.length > 1) split = split.slice(1);
    if (split.length > wanted) split = split.slice(split.length - wanted);
    return split.map((line) => line.replace(/\r+$/, ""));
  } finally {
    closeSync(handle);
  }
}

function count(buffer: Buffer, byte: number): number {
  let found = 0;
  for (const value of buffer) if (value === byte) found += 1;
  return found;
}
